File: ruby/double_linked_list/test_double_linked_list.rb
require './double_linked_list.rb'
require 'test/unit'

def fillArray(dll)
  dll.add(10)
  dll.add(20)
  dll.add(30)
end

def fillArrayRand(dll)
  # in case we accidentally generate a sorted array
  # we'll get new numbers here
  [*0...10].map { 10 + rand(89) }.each { |e| dll.insertAt(0, e) }
end

def isSorted?(dll)
  ary = dll.to_a
  last = nil
  for i in ary
    if last && i < last
      return false
    end
    last = i
  end
  true
end

class TestDoubleLinkedList < Test::Unit::TestCase

  def setup
    @dll = DoubleLinkedList.new
  end

  def teardown
    @dll = nil
  end

  def testAdd
    @dll.add(30)
    assert_equal(@dll.head.value, 30)
    @dll.add(10)
    assert_equal(@dll.head.value, 10)
  end

  def testTo_a
    assert_equal(@dll.to_a.size, 0)
    fillArray(@dll)
    assert_equal(@dll.to_a.size, 3)
  end

  def testTo_s
    assert_equal(@dll.to_s, '[]')
    fillArray(@dll)
    assert_equal(@dll.to_s, '[ 10, 20, 30 ]')
  end

  def testSwapIndexes
    fillArray(@dll)
    assert_equal(@dll.head.value, 10)
    @dll.swapIndexes(0, 2)
    assert_equal(@dll.head.value, 30)
    @dll.swapIndexes(2, 0)
    assert_equal(@dll.head.value, 10)
  end

  def testIndexOf
    fillArray(@dll)
    assert_equal(@dll.indexOf(20), 1)
    assert_equal(@dll.indexOf(-1), nil)
  end

  def testInsertAt
    fillArray(@dll)
    @dll.insertAt(0, 5)
    assert_equal(@dll.head.value, 5)
    @dll.insertAt(4, 4)
    assert_equal(@dll.tail.value, 4)
    @dll.insertAt(-1, -1)
    assert_equal(@dll.tail.value, -1)
  end

  def testClear
    fillArray(@dll)
    assert(!!@dll.head)
    @dll.clear
    assert(!@dll.head)
  end

  def testInsertSort
    fillArrayRand(@dll)
    while isSorted?(@dll) # ensure the random array isn't sorted already
      @dll.clear
      fillArrayRand(@dll)
    end

    assert(!isSorted?(@dll))
    @dll.insertSort
    assert(isSorted?(@dll))
  end

  def testShuffle
    fillArrayRand(@dll)
    @dll.insertSort
    assert(isSorted?(@dll))
    @dll.shuffle
    assert(!isSorted?(@dll))
  end

  def testBubbleSort
    fillArrayRand(@dll)
    while isSorted?(@dll) # ensure the random array isn't sorted already
      @dll.clear
      fillArrayRand(@dll)
    end

    assert(!isSorted?(@dll))
    @dll.insertSort
    assert(isSorted?(@dll))
  end

  def testLastIndex
    fillArrayRand(@dll)
    assert_equal(@dll.lastIndex, 9)
  end

  def testFind
    fillArrayRand(@dll)
    toFind = @dll.to_a.sample
    assert(!!@dll.find(toFind))
    assert(!@dll.find(-1))
  end

  def testDelete
    fillArray(@dll)
    @dll.delete(10)
    assert_equal(@dll.head.value, 20)
    @dll.delete(30)
    assert_equal(@dll.head.nxt.value, 20)
  end

  def testDeleteAt
    fillArray(@dll)
    @dll.deleteAt(0)
    assert_equal(@dll.head.value, 20)
    @dll.deleteAt(1)
    assert_equal(@dll.head.nxt.value, 20)
  end

end

File: python/double_linked_list/test_double_linked_list.py
import unittest
from double_linked_list import DoubleLinkedList
import random
from random import sample

dll = DoubleLinkedList()
randAry = None

def fillListSimple(l):
  l.addSorted(10)
  l.addSorted(20)
  l.addSorted(30)

def fillListRand(l):
  randAry = list(map(lambda x: random.randint(10, 99), range(0, 10)))
  for i in randAry:
    l.insertAt(0, i)

def verifySorted(dll):
  ary = dll.to_a()
  last = None
  for n in ary:
    if last and n < last:
      return False
    last = n
  return True

class TestSum(unittest.TestCase):

  def tearDown(self):
    dll.clear()

  def testAddSorted(self):
    dll.addSorted(10)
    self.assertEqual(dll.head.v, 10)

  def testTo_s(self):
    firstString = dll.to_s()
    self.assertEqual(firstString, '[]')

    dll.addSorted(10)
    secondString = dll.to_s()
    self.assertEqual(secondString, '[ 10 ]')

    dll.addSorted(20)
    thirdString = dll.to_s()
    self.assertEqual(thirdString, '[ 10, 20 ]')

  def testTo_a(self):
    self.assertEqual(dll.to_a(), [])

    fillListSimple(dll)
    self.assertEqual(dll.to_a(), [10,20,30])


  def testSwapIndexes(self):
    fillListSimple(dll)
    self.assertEqual(dll.head.v, 10)
    self.assertEqual(dll.head.prev.v, 30)
    dll.swapIndexes(0, 2)
    self.assertEqual(dll.head.v, 30)
    self.assertEqual(dll.head.prev.v, 10)
    dll.swapIndexes(2, 0)
    self.assertEqual(dll.head.v, 10)
    self.assertEqual(dll.head.prev.v, 30)

  def testIndexOf(self):
    fillListSimple(dll)
    self.assertEqual(dll.indexOf(20), 1)
    self.assertEqual(dll.indexOf(-1), None)

  def testInsertAt(self):
    fillListSimple(dll)
    dll.insertAt(0, 5)
    self.assertEqual(dll.head.v, 5)
    dll.insertAt(4, 4)
    self.assertEqual(dll.tail().v, 4)
    dll.insertAt(-1, -1)
    self.assertEqual(dll.tail().v, -1)

  def testClear(self):
    fillListSimple(dll)
    self.assertEqual(dll.head.v, 10)
    dll.clear()
    self.assertEqual(dll.head, None)

  def testInsertSort(self):
    fillListRand(dll)
    self.assertEqual(verifySorted(dll), False)
    dll.insertSort()
    self.assertEqual(verifySorted(dll), True)

  def testShuffle(self):
    fillListRand(dll)
    dll.insertSort()
    self.assertEqual(verifySorted(dll), True)
    dll.shuffle()
    self.assertEqual(verifySorted(dll), False)

  def testBubbleSort(self):
    fillListRand(dll)
    self.assertEqual(verifySorted(dll), False)
    dll.bubbleSort()
    self.assertEqual(verifySorted(dll), True)

  def testFind(self):
    fillListRand(dll)
    randItem = sample(dll.to_a(), 1)[0]
    self.assertEqual(dll.find(randItem) == None, False)
    self.assertEqual(dll.find(-1) == None, True)

  def testLastIndex(self):
    fillListRand(dll)
    self.assertEqual(dll.lastIndex(), 9)

  def testDelete(self):
    fillListSimple(dll)
    self.assertEqual(dll.head.v, 10)
    dll.delete(10)
    self.assertEqual(dll.head.v, 20)
    self.assertEqual(dll.head.prev.v, 30)
    dll.delete(30)
    self.assertEqual(dll.head.prev.v, 20)

  def testDeleteAt(self):
    fillListSimple(dll)
    self.assertEqual(dll.head.v, 10)
    dll.deleteAt(0)
    self.assertEqual(dll.head.v, 20)
    self.assertEqual(dll.head.prev.v, 30)
    dll.deleteAt(1)
    self.assertEqual(dll.head.prev.v, 20)




if __name__ == '__main__':
  unittest.main()